File: dayEightP1.c
#include "lib.c"
#include <stdio.h>
#include <stdlib.h>

int
get_line_size(char* str)
{
    int n = 0;

    for (int i = 0; i < strlen(str); i++) {
        if (str[i] == '\n')
            break;
        n++;
    }

    return n;
}

int
get_num_lines(char* str)
{
    int n = 0;

    for (int i = 0; i < strlen(str); i++) {
        if (str[i] == '\n')
            n++;
    }

    return n;
}

char**
get_lines(char* input, int num_lines, int line_len)
{

    char** lines = (char**)malloc(num_lines * sizeof(char*));

    char* curr_line_start = input;

    for (int i = 0; i < num_lines; i++) {
        lines[i] = (char*)malloc(line_len * sizeof(char));

        strncpy(lines[i], curr_line_start, line_len);
        curr_line_start = strchr(curr_line_start, '\n');

        curr_line_start++; // skip the whitespace
    }

    return lines;
}

typedef struct Antenna
{
    int r;
    int c;
    char freq;
} Antenna;

int min(int a, int b) {
    return (a < b) ? a : b;
}

int max(int a, int b) {
    return (a > b) ? a : b;
}


int
main()
{
    char* input = get_input_day("eight");
    // printf("%s\n", input);

    int line_size = get_line_size(input);
    int num_lines = get_num_lines(input);
    char** lines = get_lines(input, num_lines, line_size);

    size_t antenas_size = 0;
    Antenna* antenas = (Antenna*)malloc(antenas_size * sizeof(Antenna));

    for (int r = 0; r < num_lines; r++) {
        for (int c = 0; c < line_size; c++) {
            if (lines[r][c] == '.')
                continue;
            Antenna a = (Antenna){ .c = c, .r = r, .freq = lines[r][c] };

            antenas_size++;
            antenas = realloc(antenas, antenas_size * sizeof(Antenna));
            antenas[antenas_size - 1] = a;
        }
    }

    for (int i = 0; i < antenas_size; i++) {
        Antenna a1 = antenas[i];
        for (int j = 0; j < antenas_size; j++) {
            Antenna a2 = antenas[j];

            if (a1.r == a2.r && a1.c == a2.c)
                continue;

            if (a1.freq != a2.freq)
                continue;

            int row_diff = abs(a1.r - a2.r);
            int col_diff = abs(a1.c - a2.c);

            int is_top_left =
                (
                 (a1.r < a2.r && a1.c < a2.c)
                 || (a1.r > a2.r && a1.c > a2.c)
                 ) ? 1 : 0;

            int taR = min(a1.r, a2.r) - row_diff;
            int taC = ((a1.r < a2.r) ? a1.c : a2.c) + (((is_top_left) ? -1 : 1) * col_diff);
            int baR = max(a1.r, a2.r) + row_diff;
            int baC = ((a1.r > a2.r) ? a1.c : a2.c) + (((is_top_left) ? 1 : -1) * col_diff);

            if (taR >= 0 && taR < num_lines
                && taC >= 0 && taC < line_size)
                lines[taR][taC] = '#';

            if (baR >= 0 && baR < num_lines
                && baC >= 0 && baC < line_size)
                lines[baR][baC] = '#';
        }
    }
        
    int res = 0;
    for (int r = 0; r < num_lines; r++)
        // printf("%s\n", lines[r]);
        for (int c = 0; c < num_lines; c++)
        if (lines[r][c] == '#')
            res++;

    printf("%d\n", res);


    for (int i = 0; i < num_lines; i++)
        free(lines[i]);
    free(lines);
    free(input);
    return 0;
}

File: dayEightP1.php
<?php
/**
 * @var string $input
 */

$input = trim($input);

$lines = explode("\n", $input);
$anthenas = [];

for ($r = 0; $r < sizeof($lines); $r++)
{
    for ($c = 0; $c < sizeof($lines); $c++)
    {
        if ($lines[$r][$c] != '.')
        {
            $anthenas[$lines[$r][$c]][] = [$r, $c];
        }
    }
}

// echo json_encode($anthenas) . "\n";

foreach ($anthenas as $k => $points)
{
    foreach ($points as $point1)
    {
        foreach ($points as $point2)
        {
            if ($point1[0] == $point2[0] && $point1[1] == $point2[1])
                continue;

            $r1 = $point1[0];
            $c1 = $point1[1];
            $r2 = $point2[0];
            $c2 = $point2[1];

            $row_diff    = abs($r1 - $r2);
            // if the top anthena is to the left
            $is_top_left = (($r1 < $r2 && $c1 < $c2) || ($r1 > $r2 && $c1 > $c2)) ? true : false;
            $col_diff    = abs($c1 - $c2);

            // ta -> top antinode
            $taR = min($r1, $r2) - $row_diff;
            $taC = (($r1 < $r2) ? $c1 : $c2) + ((($is_top_left) ? -1 : 1) * $col_diff);
            $baR = max($r1, $r2) + $row_diff;
            $baC = (($r1 > $r2) ? $c1 : $c2) + ((($is_top_left) ? 1 : -1) * $col_diff);

            if ($taR >= 0 && $taR < sizeof($lines)
                && $taC >= 0 && $taC < strlen($lines[0]))
                $lines[$taR][$taC] = '#';

            if ($baR >= 0 && $baR < sizeof($lines)
                && $baC >= 0 && $baC < strlen($lines[0]))
                $lines[$baR][$baC] = '#';
            
        //     echo "pair: " . json_encode($point1) . " | " . json_encode($point2) . "\n";
        //     echo "taR: $taR, taC: $taC \nbaR: $baR, baC: $baC\n\n";
        }
    }
}

$lines_with_anti = implode("\n", $lines);
// echo  $lines_with_anti . "\n";


$res = 0;
for ($i = 0; $i < strlen($lines_with_anti); $i++)
{
    if ($lines_with_anti[$i] == '#')
        $res++;
}

echo $res . "\n";
